rgw: reduce per-shard entry count during ordered bucket listing

ivancich opened the pull request:
Currently, if someone requests the 1000 next entries from a bucket, each bucket index shard will receive a request for the 1000 next entries. When there are hundreds, thousands, or tens of thousands of bucket index shards, this results in a huge amplification of the request, even though only 1000 entries will be returned.

These changes reduce the per-bucket index shard requests. These also allow re-requests in edge cases where all of one shard's returned entries are consumed. Finally these changes improve the determination of whether the resulting list is truncated.

Tracker: https://tracker.ceph.com/issues/42358

mattbenjamin commented on src/rgw/rgw_rados.cc:
```diff
@@ -8007,36 +8033,67 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
+      if (0 == additional_reads--) {
+	ldout(cct, 10) << "RGWRados::" << __func__ << ": returning " <<
+	  count << "entries early as we used all " <<
+	  allowed_additional_reads << " additional reads" << dendl;
```
what happens if we hit the limit on additional reads?

ivancich replied:
This is called by RGWRados::Bucket::List::list_objects_ordered, which will re-call this function if it gets fewer than requested number of entries (and not truncated). That will refresh all shards.

Was chatting with @cbodley and he was chatting with @yehudasa who said that the S3 protocol allows fewer than what was asked for to be returned. So we could legally return whatever we were able to get.

mattbenjamin replied:
interesting--big help

markhpc commented:
@ivancich Yay!  This is going to be great.  Linking in the paper I mentioned on IRC:

http://www.dblab.ntua.gr/~gtsat/collection/scheduling/Balls%20into%20Bins%20A%20Simple%20and%20Tight%20Anal.pdf

FWIW I believe we can use the formula:

factor = ((page_size / shards) / (page_size / shards) - sqrt(2*page_size*log(shards)/shards)) - 1

To determine how many additional entries we should fetch based on the requested <page_size> (1000 often for S3) and <num_shards> to hit an expected minimum of <page_size/shards> entries per shard.

On the other hand, if we are allowed to return fewer entries than the requested page size, we could just fetch 1/<num_shards> and typically return fewer.

markhpc commented on src/rgw/rgw_rados.cc:
```diff
@@ -7913,6 +7913,21 @@ int RGWRados::cls_obj_set_bucket_tag_timeout(RGWBucketInfo& bucket_info, uint64_
+						      uint32_t num_shards)
+{
+  constexpr uint32_t preferred_min_read = 10;
+  constexpr uint32_t read_factor = 4;
```
See earlier comment about computing the expected factor

mattbenjamin replied:
if we were going to use that, it seems like we should pre-compute it

markhpc replied:
@mattbenjamin We definitely could precompute the ln(shards)/shards portion whenever we reshard, and it's probably not a terrible idea on reshard to compute/store a factor for the 1000 entry page_size case since that's what will almost always be requested.

Having said that I'm not sure it will matter much.  We currently can return anywhere from 7000-25000 bucket entries (ie 7-25 pages) per second on a single NVMe backed OSD during ordered listing (tested up to 6.4M objects, likely slower with more).  I don't think computing this per page will add much to the latency until we speed up other aspects of bucket listing.

mattbenjamin replied:
that doesn't mean we should add cpu-intensive work to the call path

cbodley replied:
if some additional cpu cycles in radosgw can save us a lot of read amp on the osds, i think it's worth doing. can we find a way to measure the read amp on some representative corpus of names, so we can compare this algorithm to some more naive ones?

markhpc replied:
@cbodley sounds like we might not need a factor at all if we can just return a truncated set of results.  I'd vote going in that direction instead since it doesn't require any book keeping or complicated math.  

Also after speaking with Sage I don't think using the min is going to work well as it doesn't scale well to large numbers of shards (the expected minimum quickly hits 0 and then you have a divide by zero to deal with).  Something like max/average as the factor might work better.

I was curious how much overhead we already have so I ran a quick RGW ordered bucket list test with perf attached to the radosgw process.  This is listing 100,000 objects from a single bucket with 16 shards (ie 16x read-amp):

|  |  |
| -- | -- |
| Entries/Page | 1000 |
| Total Pages | 100 |
| Time | 6.1s |
| Pages/second | 16.4 |
| Total RadosGW Cycles During Test | 6,408,002,175 |
| RadosGW Cycles/Page | 64,080,022 |

The takeaway for me is that we shouldn't be afraid to spend a few extra cycles/page on some math if it saves us a lot of read-amp (and more than likely would actually reduce CPU usage by reducing read-amp!).

ivancich replied:
> Also after speaking with Sage I don't think using the min is going to work well as it doesn't scale well to large numbers of shards (the expected minimum quickly hits 0 and then you have a divide by zero to deal with). Something like max/average as the factor might work better (if we can't truncate).

I'm not following this, @markhpc. If anything, having more shards than the requested entries (i.e., page size) argues for why the minimum is necessary.

markhpc replied:
@ivancich I might not have been very clear, I mean the min formula I posted earlier to use as the factor:
```
factor = ((page_size / shards) / (page_size / shards) - sqrt(2 * page_size * ln(shards)/shards))
```

As the number of shards grows without an associated growth in page size you end up with the min at or below zero (which you normally would truncate to zero) but then you end up with a divide by zero you have to solve.

Instead I'm playing with using max/average which is probably not right but so far is working well enough and still grows as the number of shards grows, but not as aggressively.

```
  float read_factor = ((1.0 * num_entries / num_shards) + sqrt(2.0 * num_entries*log(num_shards)/num_shards)) / (1.0 * num_entries / num_shards);
```

markhpc replied:
Ran some new tests with 64 buckets and 6.4M objects:

|  | Master | #30853 | #30853 + max_expr(1) |
| -- | -- | -- | -- |
| Entries/Page | 1000 | 1000 | 1000 |
| Total Pages | 6400 | 6400 | 6400 |
| Time | 778.0s | 288.7s | 272.9s |
| Pages/second | 8.23 | 22.17 | 23.45 |
| Total RadosGW Cycles During Test | 1,631,302,890,773 | 274,418,770,960 | 213,295,217,820 |
| RadosGW Cycles/Page | 254,891,077 | 42,877,933 | 33,327,379 |

mattbenjamin replied:
@markhpc sure, but log(shards) at the very least is for the given bucket a constant.  Throw me a bone here.

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -8007,36 +8033,67 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
+      // corresponding to oids, and we only have one entry there
+      r = CLSRGWIssueBucketList(ioctx, name, prefix, num_entries_per_shard,
+				list_versions, shard_oid, list_results,
+				cct->_conf->rgw_bucket_index_max_aio)();
```
if you're only fetching a single shard, can you use `cls_rgw_bucket_list_op()` instead of `CLSRGWIssueBucketList` the way `cls_bucket_list_unordered()` does?

cbodley replied:
p.s. i'm skeptical about this part that fetches more from single shards, because we don't get any parallelism from these requests. we may be better off just returning a truncated listing to the client, and resuming the parallel listing on their next request

i feel like this part is orthogonal to the change that limits how many entries we request from each shard - can we investigate this in a separate pr?

mattbenjamin replied:
I would tend to agree--given the earlier proof that clients will send further requests

markhpc replied:
If we are allowed to return truncated listings (and there aren't lots of misbehaving clients), it seems like that might be the way to go.   IE fetch 1/<num_shards> entries from each shard and return a mostly complete listing with no extra book keeping or re-fetching.

ivancich replied:
So if we choose our "excess" per-shard # of entries well, per shard re-fetches should be very rare. When we do hit the end of one shard, if reading one shard a few times saves us from re-issuing reads to 2,000 shards (or 65,000 shards, which we allow), I think it would be worth it. Because if we return fewer than the requested # of entries, we're pretty much guaranteed that either the ultimate client or the internal list op will re-issue, hitting all shards.

So my proposal is to limit the # of per-shard re-reads to a small #, such as 2. I would also like to log these events at level 1, so we can get some data. Again, the re-reads and log entries should be rare. I know @cbodley is disinclined to to log extraneous stuff at level 1, but it can be labelled "INFO" so as not to raise concern from end-users.

ivancich replied:
@cbodley wrote:
> we may be better off just returning a truncated listing to the client, and resuming the parallel listing on their next request

My thinking is starting to change towards this view, although I'd like to discuss more.

So the question is whether the parameter should be interpreted as # of desired entries or maximum # of entries the client can handle at a time.

The more typical use case is the desire to list the entire bucket, in which case no matter what's returned, it will be followed by additional requests until the end of the bucket list is reached. In that case the parameter is better interpreted the max that the client can handle at a time, in which case returning fewer is not a problem.

There may be use cases where the client wants *exactly* that number of items, in which case not returning that number guarantees a second call, but possibly no more. That seems wasteful in that it essentially doubles the work.

So now I'm thinking that adding padding of 25%-50% might be sufficient. I'll bet it returns what's asked for most of the time, but prevents us from adding a second call and therefore doubling the work in the second use case.

mattbenjamin replied:
I understood we'd confirmed that Amazon does not promise to return exactly any number of items, and even if it wanted to, there would be scenarios where it could not

markhpc commented:
@ivancich Looks like we do take a hit by fetching additional pages in the latest version.  We fetch 8852 pages instead of 6400 pages for 6.4M objects with an overall duration of 317.1s.  I'm changing the table to look at Entries/Second and Cycles/Second now since pages can now be of variable size.

|  | Master | #30853 old | #30853 new | #30853 new + max_expr2(1) |
| -- | -- | -- | -- | -- |
| Avg Entries/Page | 1000 | 1000 | 723 | 995 |
| Total Pages | 6400 | 6400 | 8852 | 6432 |
| Time (seconds) | 778.0 | 288.7 | 317.1 | <b>272.7</b> |
| Entries/second | 8226 | 22168 | 20183 | <b>23469</b> |
| Total RadosGW Cycles | 1,631M | 274M | <b>215M</b> | 224M |
| RadosGW Cycles/Entry | 254891 | 42878 | <b>33637</b> | 34963 |

(1) I realized I was being dumb and should have just used the max_expr directly as the return value in calc_ordered_bucket_list_per_shard (rounded up to the nearest int).  I believe this is the correct implementation of the estimation (note how it's close to but not quite hitting 1000 entries/page):

```
uint32_t RGWRados::calc_ordered_bucket_list_per_shard(uint32_t num_entries, uint32_t num_shards)
{
  return static_cast<uint32_t>((1.0 * num_entries / num_shards) + sqrt(2.0 * num_entries*log(num_shards)/num_shards) + 1);
}
```

ivancich commented:
@markhpc That's interesting and highlights some trade-offs. At 64 shards, we're playing user experience (results/second) against rgw cycles. Then at 256 shards we're playing user experience & rgw cycles against osd cycles (read amplification).

It would be interesting to see how tweaking # of shards to nearest prime value (e.g., 61, 257) would change things. 

Also, it would be interesting to see data at 1999 shards (nearest prime to 2000).

How many objects are in the test bucket?

markhpc commented:
> @markhpc That's interesting and highlights some trade-offs. At 64 shards, we're playing user experience (results/second) against rgw cycles. Then at 256 shards we're playing user experience & rgw cycles against osd cycles (read amplification).
> 
> It would be interesting to see how tweaking # of shards to nearest prime value (e.g., 61, 257) would change things.
> 
> Also, it would be interesting to see data at 1999 shards (nearest prime to 2000).
> 
> How many objects are in the test bucket?

Oh, I forgot about the min value of 8, I'll update the chart

markhpc commented:
@ivancich I did some debugging over the weekend and discovered that the increase in pages was due to how vcurrents processing in RGWRados::cls_bucket_list_ordered was changed.  By giving up whenever any shard hits 0, we end up really drawing out the final listing, sometimes only returning 1 or 2 entries per page depending on how quickly we hit that condition.  I believe we can avoid this by keeping track of when the number of entries returned for a given shard via CLSRGWIssueBucketList is below the num_entries_per_shard count.  IE if we know that there are no more entries to retrieve from that shard, we can safely remove it from the candidates and carry on much like the original code did.

I created a PR for this change and for the max_expr change here: https://github.com/ivancich/ceph-fork/pull/1  Updated numbers for 1024 shards using this PR:

| 1024 Shards, 6.4M Objects | Master | #30853 old | #30853 new | #30853 new + https://github.com/ivancich/ceph-fork/pull/1 |
| -- | -- | -- | -- | -- |
| Avg Entries/Page | 1000 | 1000| 927 | 1000 |
| Total Pages | 6400 | 6400 | 6907 | 6400 |
| Read-Amp | 1024 | 4 | ~8.84(2)(3) | <b>~4.77(2)</b> |
| Time (seconds) | 7469.9 | ? | 513.0 | <b>431.2</b> |
| Entries/second | 857 | ? | 12476 | <b>14842</b> |
| Total RadosGW Cycles | 21,354B | ? | 1,104B | <b>891B</b> |
| RadosGW Cycles/Entry | 3336582 | ? | 172484 | <b>139161</b> |

ivancich commented:
> @ivancich I did some debugging over the weekend and discovered that the increase in pages was due to how vcurrents processing in RGWRados::cls_bucket_list_ordered was changed. By giving up whenever any shard hits 0, we end up really drawing out the final phase of listing, sometimes only returning 1 or 2 entries per page depending on how quickly we hit that condition. I believe we can avoid this by keeping track of when the number of entries returned for a given shard via CLSRGWIssueBucketList is below num_entries_per_shard. IE if we know that there are no more entries to retrieve for that shard beyond what we already have, we can safely remove it from candidates and carry on much like the original code did.

This is a good catch, @markhpc . In fact the earlier version of the code that re-read shards had this logic in it, but when I dropped that, it got lost -- my fault.

I've updated my PR to examine the `is_truncated` flag in the shard's result read result. Checking the size of the result will work in most cases but would fail in the case when there happened to be the "right" number of entries at the end of the shard's list. So `is_truncated` is the way to go.

markhpc commented:
> > @ivancich I did some debugging over the weekend and discovered that the increase in pages was due to how vcurrents processing in RGWRados::cls_bucket_list_ordered was changed. By giving up whenever any shard hits 0, we end up really drawing out the final phase of listing, sometimes only returning 1 or 2 entries per page depending on how quickly we hit that condition. I believe we can avoid this by keeping track of when the number of entries returned for a given shard via CLSRGWIssueBucketList is below num_entries_per_shard. IE if we know that there are no more entries to retrieve for that shard beyond what we already have, we can safely remove it from candidates and carry on much like the original code did.
> 
> This is a good catch, @markhpc . In fact the earlier version of the code that re-read shards had this logic in it, but when I dropped that, it got lost -- my fault.
> 
> I've updated my PR to examine the `is_truncated` flag in the shard's result read result. Checking the size of the result will work in most cases but would fail in the case when there happened to be the "right" number of entries at the end of the shard's list. So `is_truncated` is the way to go.

I don't think it fails so long as we assume that getting the "right" number of entries puts us in the "maybe" state and we break?

ivancich commented:
> > > @ivancich I did some debugging over the weekend and discovered that the increase in pages was due to how vcurrents processing in RGWRados::cls_bucket_list_ordered was changed. By giving up whenever any shard hits 0, we end up really drawing out the final phase of listing, sometimes only returning 1 or 2 entries per page depending on how quickly we hit that condition. I believe we can avoid this by keeping track of when the number of entries returned for a given shard via CLSRGWIssueBucketList is below num_entries_per_shard. IE if we know that there are no more entries to retrieve for that shard beyond what we already have, we can safely remove it from candidates and carry on much like the original code did.
> > 
> > 
> > This is a good catch, @markhpc . In fact the earlier version of the code that re-read shards had this logic in it, but when I dropped that, it got lost -- my fault.
> > I've updated my PR to examine the `is_truncated` flag in the shard's result read result. Checking the size of the result will work in most cases but would fail in the case when there happened to be the "right" number of entries at the end of the shard's list. So `is_truncated` is the way to go.
> 
> I don't think it fails so long as we assume that getting the "right" number of entries puts us in the "maybe" state and we break?

Well it could "fail" in the same way my previous code "failed" (although w/ much much lower odds) -- it would break out of the loop when it does not have to and return fewer entries than it could, introducing a performance inefficiency. Fortunately we have the `is_truncated` flag in the result.

markhpc commented:
> > > > @ivancich I did some debugging over the weekend and discovered that the increase in pages was due to how vcurrents processing in RGWRados::cls_bucket_list_ordered was changed. By giving up whenever any shard hits 0, we end up really drawing out the final phase of listing, sometimes only returning 1 or 2 entries per page depending on how quickly we hit that condition. I believe we can avoid this by keeping track of when the number of entries returned for a given shard via CLSRGWIssueBucketList is below num_entries_per_shard. IE if we know that there are no more entries to retrieve for that shard beyond what we already have, we can safely remove it from candidates and carry on much like the original code did.
> > > 
> > > 
> > > This is a good catch, @markhpc . In fact the earlier version of the code that re-read shards had this logic in it, but when I dropped that, it got lost -- my fault.
> > > I've updated my PR to examine the `is_truncated` flag in the shard's result read result. Checking the size of the result will work in most cases but would fail in the case when there happened to be the "right" number of entries at the end of the shard's list. So `is_truncated` is the way to go.
> > 
> > 
> > I don't think it fails so long as we assume that getting the "right" number of entries puts us in the "maybe" state and we break?
> 
> Well it could "fail" in the same way my previous code "failed" (although w/ much much lower odds) -- it would break out of the loop when it does not have to and return fewer entries than it could, introducing a performance inefficiency. Fortunately we have the `is_truncated` flag in the result.

Ah, that would make sense to use the is_truncated flag for the same purpose.  Does it actually set is_truncated to to true if it was successfully able to return the full list of entries but no more are available?  I haven't dug into that portion of the CLS code yet.

ivancich commented:
> Ah, that would make sense to use the is_truncated flag for the same purpose. Does it actually set is_truncated to to true if it was successfully able to return the full list of entries but no more are available? I haven't dug into that portion of the CLS code yet.

Yeah, that's the role of that flag.

markhpc commented:
@ivancich Your new version segfaulted near the end of the bucket listing test (page 6394).  here's a BT of thread 1 from gdb:

```
Thread 1 (Thread 0x7fe4b3adf700 (LWP 95999)):
#0  0x00007fe4cdd11c1f in raise () from /lib64/libpthread.so.0
#1  0x000055a449f5bd46 in reraise_fatal (signum=11) at /home/nhm/src/ceph/ceph/src/global/signal_handler.cc:326
#2  handle_fatal_signal (signum=11) at /home/nhm/src/ceph/ceph/src/global/signal_handler.cc:326
#3  <signal handler called>
#4  0x000055a44a04856e in rgw_obj_key::parse_index_key (ns=0x55a452cd1a20, name=0x55a452cd19e0, key="") at /usr/include/c++/8/bits/basic_string.h:1037
#5  rgw_obj_key::rgw_obj_key (this=0x55a452cd19e0, k=...) at /home/nhm/src/ceph/ceph/src/rgw/rgw_common.h:1694
#6  0x000055a44a246ca1 in rgw_obj::rgw_obj (k=..., b=..., this=0x55a452cd18a0) at /home/nhm/src/ceph/ceph/src/rgw/rgw_common.h:2230
#7  RGWRados::check_disk_state (this=0x55a44ce67b00, io_ctx=..., bucket_info=..., list_state=..., object=..., suggested_updates=..., y=...) at /home/nhm/src/ceph/ceph/src/rgw/rgw_rados.cc:8378
#8  0x000055a44a252eb6 in RGWRados::cls_bucket_list_ordered (this=<optimized out>, bucket_info=..., shard_id=<optimized out>, start=..., prefix=..., num_entries=<optimized out>, list_versions=<optimized out>, m=..., is_truncated=<optimized out>, last_entry=<optimized out>, y=..., force_check_filter=<optimized out>)
    at /usr/include/c++/8/bits/stl_vector.h:930
#9  0x000055a44a254871 in RGWRados::Bucket::List::list_objects_ordered (this=this@entry=0x55a452cd2cb0, max_p=<optimized out>, result=<optimized out>, common_prefixes=<optimized out>, is_truncated=<optimized out>, y=...) at /home/nhm/src/ceph/ceph/src/rgw/rgw_rados.h:879
#10 0x000055a44a1ce50c in RGWRados::Bucket::List::list_objects (y=..., is_truncated=0x55a44db7bf2c, common_prefixes=0x55a44db7bef8, result=0x55a44db7bee0, max=<optimized out>, this=0x55a452cd2cb0) at /home/nhm/src/ceph/ceph/src/rgw/rgw_rados.h:1009
#11 RGWListBucket::execute (this=0x55a44db7bb00) at /home/nhm/src/ceph/ceph/src/rgw/rgw_op.cc:2815
#12 0x000055a449f79b04 in rgw_process_authenticated (handler=<optimized out>, op=@0x55a452cd3400: 0x55a44db7bb00, req=0x55a452cd4910, s=0x55a452cd38b0, skip_retarget=<optimized out>) at /home/nhm/src/ceph/ceph/src/rgw/rgw_process.cc:161
#13 0x000055a449f7cee0 in process_request (store=0x55a44cd22a20, rest=0x7ffc8cc7ae00, req=req@entry=0x55a452cd4910, frontend_prefix="", auth_registry=..., client_io=0x55a452cd4930, olog=0x0, yield=..., scheduler=0x55a44daa1148, http_ret=0x0) at /home/nhm/src/ceph/ceph/src/rgw/rgw_process.cc:278
--Type <RET> for more, q to quit, c to continue without paging--
#14 0x000055a449eeb244 in (anonymous namespace)::handle_connection<boost::asio::basic_stream_socket<boost::asio::ip::tcp> > (context=..., env=..., stream=..., buffer=..., pause_mutex=..., scheduler=<optimized out>, ec=..., yield=..., is_ssl=false) at /usr/include/c++/8/bits/shared_ptr_base.h:996
#15 0x000055a449eebfb4 in (anonymous namespace)::AsioFrontend::<lambda(boost::asio::yield_context)>::operator() (yield=..., __closure=0x55a4500e58f8) at /usr/include/c++/8/bits/unique_ptr.h:342
#16 boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::executor_type> >, (anonymous namespace)::AsioFrontend::accept((anonymous namespace)::AsioFrontend::Listener&, boost::system::error_code)::<lambda(boost::asio::yield_context)> >::operator() (
    ca=..., this=<optimized out>) at /home/nhm/src/ceph/ceph/build/boost/include/boost/asio/impl/spawn.hpp:382
#17 boost::coroutines::detail::push_coroutine_object<boost::coroutines::pull_coroutine<void>, void, boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::executor_type> >, (anonymous namespace)::AsioFrontend::accept((anonymous namespace)::AsioFrontend::Listener&, boost::system::error_code)::<lambda(boost::asio::yield_context)> >&, boost::coroutines::basic_standard_stack_allocator<boost::coroutines::stack_traits> >::run (this=0x55a452cd5f60) at /home/nhm/src/ceph/ceph/build/boost/include/boost/coroutine/detail/push_coroutine_object.hpp:293
#18 boost::coroutines::detail::trampoline_push_void<boost::coroutines::detail::push_coroutine_object<boost::coroutines::pull_coroutine<void>, void, boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::executor_type> >, (anonymous namespace)::AsioFrontend::accept((anonymous namespace)::AsioFrontend::Listener&, boost::system::error_code)::<lambda(boost::asio::yield_context)> >&, boost::coroutines::basic_standard_stack_allocator<boost::coroutines::stack_traits> > >(boost::context::detail::transfer_t) (t=...)
    at /home/nhm/src/ceph/ceph/build/boost/include/boost/coroutine/detail/trampoline_push.hpp:70
#19 0x000055a44a400c8f in make_fcontext ()
#20 0x000055a44a7969d0 in vtable for boost::coroutines::detail::push_coroutine_object<boost::coroutines::pull_coroutine<void>, void, boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::executor_type> >, (anonymous namespace)::AsioFrontend::accept((anonymous namespace)::AsioFrontend::Listener&, boost::system::error_code)::{lambda(boost::asio::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::executor> >)#4}>&, boost::coroutines::basic_standard_stack_allocator<boost::coroutines::stack_traits> > ()
#21 0x0000000000000026 in ?? ()
#22 0x0000000000000000 in ?? ()
```

ivancich commented:
> @ivancich Your new version segfaulted near the end of the bucket listing test (page 6394). I actually ran the test twice and it appears to have died roughly in the same place, but with a different backtrace. here's a BT of thread 1 from gdb:

Yeah, that's fixed now, @markhpc .

ivancich commented:
The current version passes s3-tests. Will next run it against teuthology rgw suite.

markhpc commented:
@ivancich yep, your latest push fixed it.  Here are updated numbers for the 1024 shard case.  I think we still likely want https://github.com/ivancich/ceph-fork/pull/1/commits/9cb886ce866a8832ce9f30d6234fa3bafe651e54 even if we don't want to use the first commit.  I also have a couple of other minor changes that might help slightly.

| 1024 Shards, 6.4M Objects | Master | #30853 v1 | #30853 v2 | #30853 v2 + https://github.com/ivancich/ceph-fork/pull/1 |
| -- | -- | -- | -- | -- |
| Avg Entries/Page | 1000 | 927 | 1000 | 1000 |
| Total Pages | 6400  | 6907 | 6400 | 6400 |
| Read-Amp | 1024 | ~8.84(1)(2) | <b>~5.12(1)</b> | 8.192
| Time (seconds) | 7469.9 | 513.0 | <b>431.2</b> | 473.4
| Entries/second | 857 | 12476 | <b>14842</b> | 13519
| Total RadosGW Cycles | 21,354B | 1,104B | <b>891B</b> | 1,061B
| RadosGW Cycles/Entry | 3,336,582 | 172,484 | <b>139,161</b> | 165,741

cbodley commented:
thanks for putting these numbers together @markhpc, i didn't expect the gains to be this massive!

markhpc commented:
@cbodley I think we still have some gains to be realized in cls_bucket_list_ordered, though the lowest hanging fruit was definitely avoiding the massive read amp with large buckets.  I'm not sure we can get the read amp much lower than what @ivancich's PR + my second commit in https://github.com/ivancich/ceph-fork/pull/1 does so we'll now have to focus on reducing latency and cpu cycles / op via other routes now.  Still, 14842 IOPS isn't terrible.  It's about 60-65% of the speed I was seeing previously in single-shard tests and even closer to single-shard performance in previous releases prior to some of the other performance improvements.

mattbenjamin commented:
@markhpc what's been happening with rocksdb sharding?

markhpc commented:
> @markhpc what's been happening with rocksdb sharding?

I ran through some tests of Adam's PR last week.  The gist of it is that we see a large reduction in compaction overhead but no real performance differences in either the small RGW or RBD workloads we tested as compaction was not the primary bottleneck.  You can see the results here:

https://docs.google.com/spreadsheets/d/1FYFBxwvE1i28AKoLyqrksHptE1Z523NU3Fag0MELTQo/edit?usp=sharing

ivancich commented:
> @cbodley I think we still have some gains to be realized in cls_bucket_list_ordered, though the lowest hanging fruit was definitely avoiding the massive read amp with large shard counts. I'm not sure we can get the read amp much lower than what @ivancich's PR + my second commit in [ivancich#1](https://github.com/ivancich/ceph-fork/pull/1) does so we'll now have to focus on reducing latency and cpu cycles / op via other routes. Still, ~15K entries/second at 1024 shards isn't terrible. It's about 60-65% of the speed I was seeing in recent master single-shard tests and even closer to single-shard performance in previous releases prior to some of the other performance improvements.

As you know, @markhpc, there is an as yet unmerged PR (https://github.com/ceph/ceph/pull/30272) that moves the filtering to the osd/cls, when a delimiter is specified, which I think is every S3 bucket listing. That should also help in reducing what is returned from osd to rgw in many cases.

markhpc commented:
> > @cbodley I think we still have some gains to be realized in cls_bucket_list_ordered, though the lowest hanging fruit was definitely avoiding the massive read amp with large shard counts. I'm not sure we can get the read amp much lower than what @ivancich's PR + my second commit in [ivancich#1](https://github.com/ivancich/ceph-fork/pull/1) does so we'll now have to focus on reducing latency and cpu cycles / op via other routes. Still, ~15K entries/second at 1024 shards isn't terrible. It's about 60-65% of the speed I was seeing in recent master single-shard tests and even closer to single-shard performance in previous releases prior to some of the other performance improvements.
> 
> As you know, @markhpc, there is an as yet unmerged PR (#30272) that moves the filtering to the osd/cls, when a delimiter is specified, which I think is every S3 bucket listing. That should also help in reducing what is returned from osd to rgw in many cases.

Would you expect a reduction in these tests though where we are listing the entire bucket contents?  I had assumed that in both cases we wouldn't be doing any filtering in this particular test.

cbodley commented:
@ivancich what's left to do here?

markhpc commented:
> @ivancich what's left to do here?

@cbodley @ivancich It would be really nice if we could include https://github.com/ivancich/ceph-fork/pull/1/commits/9cb886ce866a8832ce9f30d6234fa3bafe651e54 so it gets tested in the same run.  I can submit it as a second PR but we'll just have to test it all over again and I think the results from above provide reasonable evidence that it's helping.

cbodley commented:
@markhpc i agree, yeah. i was particularly impressed that 'balls into bins' was able to return all 1000 entries on average (meaning that barely any returned fewer), while also lowering read-amp compared to this pr

ivancich commented:
Thanks, everyone -- @mattbenjamin @markhpc @cbodley! These discussions have been incredibly interesting. I'd like to lock in these gains and wait until a more convenient time to do further work (e.g., memoization). But you, the reviewers, will have the final say!

ivancich commented:
> Read-Amp 	1024 	~8.84(1)(2) 	~5.12(1) 	8.192

The 8.192 looked a little too interesting. In discussing w/ @cbodley we realized that it was because that version of the PR read a minimum of 8 entries per shard. And the test had 1024 shards and a page size of 1000. 8 * 1024 / 1000 = 8.192!!

The idea of having a minimum is for when num_shards >> (much greater) page_size, which could result in reading only 1 entry per shard. Once we take the earliest entry among the shards and exhausted its that shard's supply, we do not know if we'd have needed the next entry to complete our page size. So we'd be stuck. The other side of that coin is that it's not much additional work for an osd to return a small number of entries over just one entry given the overheads involved (network, request decoding, bucket-index-shard reading, etc.).

So that minimum is now at 4.

markhpc commented:
@ivancich 4 is probably better than 8, but now you'll have a different problem where for any sufficiently large number of shards you may not return the full number of entries leading to worst performance as the number of shards grow.  I think that's why we want an adaptive solution like in https://github.com/ivancich/ceph-fork/commit/9cb886ce866a8832ce9f30d6234fa3bafe651e54

ivancich commented:
@markhpc Unless I'm misunderstanding you, it sounds like you may have missed the latest code update in the PR in my force push from yesterday....

markhpc commented:
@ivancich ah, I was looking for the commit, I didn't realize you had copied the code directly into your commit.  I don't think there's much value in specifying a min anymore.  You'll always fetch at least 1 record and I don't think you want to fetch 4 in situations where it specifies fewer.  Is there something I'm missing?

ivancich commented:
@markhpc The issue with only returning one entry per shard is -- if that shard is also marked truncated, then as soon we consume the entry from that shard we can add no more results to the page returned since we cannot know that the next entry would not come from that shard (and we haven't read it and we don't do re-reads).

Given the necessary overheads (network traffic, decoding the request, etc.) of the calls to the OSDs, returning 4 entries from each shard is not that much more work than returning 1 entry from each shard. So that's the argument for a minimum, whether it be 4 or 8.

The numbers you show are interesting. They could argue for just using a minimum. Just looking at the last three tables, requesting a minimum of 8 would work well in all cases (in the first two tables, 8 would likely never be returned, but we're just asking for 8).

Do I understand your last table correctly in that in the 1000 entries / 1024 shards case, there's a 30+% chance that one shard will have more than 4 entries? If that's the correct interpretation, I find that very counter-intuitive.

Also I don't know if your tables account for this, but just as specified in the first paragraph of this comment, the number of entries we return is determined by the point at which we exhaust the entries for a shard that's marked as truncated, just because we do not know if any subsequent entry we need might come from that shard. That's a different stop trigger than checking whether the first n entries (n = page size) were included in what was returned.

markhpc commented:
@ivancich 

The test program tries to simulate the case where we set the number of entries per shard upfront and then randomly selects from them until any given shard hits 0.  So in the 4 entries_per_shard case, we stopped after 686 entries were reached out of 1000 on average over 100K test runs.

ivancich commented:
> The test program tries to simulate the case where we set the number of entries per shard upfront and then randomly selects from them until any given shard hits 0. So in the 4 entries_per_shard case, we stopped after 686 entries were reached out of 1000 on average over 100K test runs.

OK, so let me re-state my question. Would you say that's the same as saying in that in the 1000 entries / 1024 shards case, there's a ~30% chance that one shard (out of 1024) will be selected 4 times? That still is not intuitive to me.

And that's not exactly our stop criteria for filling up the return list. We'd have to select 4 items from one shard *and* there would have to be a 5th item on that shard that we did not read (i.e., the shard was truncated). If the got 4 items from that shard and it was not truncated, we can continue.

markhpc commented:
> > The test program tries to simulate the case where we set the number of entries per shard upfront and then randomly selects from them until any given shard hits 0. So in the 4 entries_per_shard case, we stopped after 686 entries were reached out of 1000 on average over 100K test runs.
> 
> OK, so let me re-state my question. Would you say that's the same as saying in that in the 1000 entries / 1024 shards case, there's a ~30% chance that one shard (out of 1024) will be selected 4 times? That still is not intuitive to me.
> 

AFAIK it's quite hard to actually compute what that probability is. I spent a fair amount of time seeing if I could answer questions like that using combinatorics but ultimately failed. That's why I ended up just writing a simulator. The simulator works by storing K-size counters in a N-length vector and then randomly decrements those counters until we either try to decrement a counter that already is 0 or we've hit the specified number of tries (1000 in this case). I believe that should give us an approximation of what would happen during bucket listing when we fetch a static number of entries per shard but entries with specific ordering positions are distributed randomly across shards.

> And that's not exactly our stop criteria for filling up the return list. We'd have to select 4 items from one shard _and_ there would have to be a 5th item on that shard that we did not read (i.e., the shard was truncated). If the got 4 items from that shard and it was not truncated, we can continue.

That was basically the goal. IE in balls-in-bins terminology, if we have N bins that each contain K balls, how many times can we randomly draw balls from bins until we attempt to draw from an empty bin?

ivancich commented:
@markhpc I realize you said in the performance meeting that you'd like to revisit some of this when your time permits. It seems best to get these benefits into master and further refinements/improvements can be future prs. Do you have a strong objection to that? I truly appreciate the knowledge and results you brought to this pr and bring to ceph!

ivancich commented:
teuthology rgw/verify: http://pulpito.ceph.com/ivancich-2019-11-09_20:41:15-rgw-wip-eric-testing-1-distro-basic-smithi/
teuthology rgw/(non-)verify: http://pulpito.ceph.com/ivancich-2019-11-09_20:40:17-rgw-wip-eric-testing-1-distro-basic-smithi/

ivancich commented:
s3-tests shows no regressions

markhpc commented:
Hi ivancich,

Sorry for the late reply!  I was just going to recommend that you merge it and if we can make further improvements we just do it via follow up PRs.  I did run through some subsequent tests, but it got bogged down once we observed the DeleteRange regression in the OSD.  The good news is that this PR as-is works fairly well but does appear to introduce some extra client side page requests as-is (this is with 1M objects and 7 shards):

```
Loop,Inteval,Duration(s),Mode,Ops,MB/s,IO/s,Min Latency (ms),Avg Latency(ms),99% Latency(ms),Max Latency(ms),Slowdowns
0,TOTAL,43.69,LIST,1218,0.00,27.88,24.57,35.87,44.61,53.38,0
```

Optimally we'd only have requested 1000 pages there instead of 1218.  I'll try to submit a follow-up PR that I believe will reduce that based on the discussion above.

Mark
